fearless/Modules/Staking/SelectValidatorsFlow/CustomValidatorList/CustomValidatorListPresenter.swift

```diff
@@ -1,46 +1,231 @@
+        localizationManager: LocalizationManagerProtocol,
+        electedValidators: [ElectedValidatorInfo],
+        recommendedValidators: [ElectedValidatorInfo],
+        maxTargets: Int,
+        logger: LoggerProtocol? = nil
     ) {
         self.interactor = interactor
         self.wireframe = wireframe
         self.viewModelFactory = viewModelFactory
-        selectedValidators = validators
+        electedValidatorList = electedValidators
+        recommendedValidatorList = recommendedValidators
+        self.maxTargets = maxTargets
+        self.logger = logger
+        self.localizationManager = localizationManager
+    }
+
+    // MARK: - Private functions
+
+    private func composeFilteredValidatorList() -> [ElectedValidatorInfo] {
+        let composer = CustomValidatorListComposer(filter: filter)
+        return composer.compose(from: electedValidatorList)
+    }
+
+    private func updateFilteredValidatorsList() {
+        filteredValidatorList = composeFilteredValidatorList()
+    }
+
+    private func provideValidatorListViewModel() {
+        let viewModel = viewModelFactory.createViewModel(
+            from: filteredValidatorList,
+            selectedValidators: selectedValidatorSet,
+            totalValidatorsCount: electedValidatorList.count,
+            filter: filter,
+            priceData: priceData,
+            locale: selectedLocale
+        )
+
+        self.viewModel = viewModel
+        view?.reload(viewModel)
+    }
+
+    private func provideFilterButtonViewModel() {
+        let emptyFilter = CustomValidatorListFilter.defaultFilter()
+        let appliedState = filter != emptyFilter
+
+        view?.setFilterAppliedState(to: appliedState)
+    }
+
+    private func provideViewModels() {
+        updateFilteredValidatorsList()
+
+        provideValidatorListViewModel()
+        provideFilterButtonViewModel()
+    }
+
+    private func performDeselect() {
+        guard var viewModel = viewModel else { return }
+
+        var indexes: [Int] = []
+
+        viewModel.cellViewModels =
+            viewModel.cellViewModels.enumerated().map { index, item in
+                var newItem = item
+
+                if newItem.isSelected {
+                    newItem.isSelected = false
+                    indexes.append(index)
+                }
+
+                return newItem
+            }
+
+        selectedValidatorSet = []
+
+        self.viewModel = viewModel
+
+        view?.reload(viewModel, at: indexes)
     }
 }
 
+// MARK: - CustomValidatorListPresenterProtocol
+
 extension CustomValidatorListPresenter: CustomValidatorListPresenterProtocol {
     func setup() {
-        let viewModel = viewModelFactory.createViewModel(validators: selectedValidators)
-        view?.reload(with: viewModel)
+        provideViewModels()
+        interactor.setup()
+    }
+
+    // MARK: - Header actions
+
+    func fillWithRecommended() {
+        var index = 0
+
+        while index < recommendedValidatorList.count,
```

Comment:
Looks like we could rewrite more concisely in functional way like: ```recommendedValidatorList.filter(...).prefix(maxTargets)```

Reply:
Well, let's try